app/services/title_builder.rb

```diff
@@ -4,38 +4,130 @@ class TitleBuilder
   # @param [Array<Cocina::Models::Title>] titles
   # @returns [String] the title value for Solr
   def self.build(titles)
-    title = titles.first
+    title = primary_title(titles) || first_untyped_title(titles) || titles.first
     if title.value
-      title.value
+      my_title = title.value
     elsif title.structuredValue
-      build_structured(title.structuredValue)
+      my_title = title_from_structured_values(title.structuredValue, non_sorting_char_count(title))
     elsif title.parallelValue
-      build(title.parallelValue)
+      my_title = build(title.parallelValue)
     end
+    remove_trailing_punctuation(my_title.strip) if my_title.present?
   end
 
-  def self.build_structured(titles)
-    title_parts = titles.map do |title|
-      if title.structuredValue
-        [build_structured(title.structuredValue), '']
+  # rubocop:disable Metrics/BlockLength
+  # rubocop:disable Metrics/CyclomaticComplexity
+  # rubocop:disable Metrics/MethodLength
+  # rubocop:disable Metrics/PerceivedComplexity
+  def self.title_from_structured_values(structured_values, non_sorting_char_count)
+    structured_title = ''
+    title_from_part = ''
+    structured_values.each do |structured_value|
+      # There can be a structuredValue inside a structuredValue.  For example,
+      #   a uniform title where both the name and the title have internal StructuredValue
+      return title_from_structured_values(structured_value.structuredValue, non_sorting_char_count) if structured_value.structuredValue
+
+      value = structured_value.value&.strip
+      next unless value
+
+      # additional types:  name, uniform ...
+      case structured_value.type&.downcase
```

Comment:
the `type&.downcase` implies that this might be `nil`, but there is no when or else case for that value. Is that something we need to deal with?

Reply:
I don't know;  Arcadia didn't have an example for it.  I asked her what other types will show up here and she is looking into it.  No reason to hold up this PR for those edge cases.